entity-service-change-event-generator/src/main/java/org/hypertrace/entity/service/change/event/impl/EntityChangeEventGeneratorImpl.java

```diff
@@ -0,0 +1,123 @@
+    entityChangeEventProducer =
+        eventStore.createProducer(
+            ENTITY_CHANGE_EVENTS_TOPIC,
+            new EventProducerConfig(
+                storeType, config.getConfig(ENTITY_CHANGE_EVENTS_PRODUCER_CONFIG)));
+  }
+
+  @Override
+  public void sendCreateNotification(Collection<Entity> entities) {
+    entities.forEach(this::sendCreateNotification);
+  }
+
+  @Override
+  public void sendDeleteNotification(Collection<Entity> entities) {
+    entities.forEach(this::sendDeleteNotification);
+  }
+
+  @Override
+  public void sendChangeNotification(
+      Collection<Entity> existingEntities, Collection<Entity> updatedEntities) {
+    Map<String, Entity> existingEntityMap =
+        existingEntities.stream().collect(Collectors.toMap(Entity::getEntityId, identity()));
+    Map<String, Entity> upsertedEntityMap =
+        updatedEntities.stream().collect(Collectors.toMap(Entity::getEntityId, identity()));
+    MapDifference<String, Entity> mapDifference =
+        Maps.difference(existingEntityMap, upsertedEntityMap);
+
+    mapDifference
+        .entriesOnlyOnRight()
+        .entrySet()
+        .forEach(
+            entry -> {
+              sendCreateNotification(entry.getValue());
+            });
+
+    mapDifference
+        .entriesDiffering()
+        .entrySet()
+        .forEach(
+            entry -> {
+              MapDifference.ValueDifference<Entity> valueDifference = entry.getValue();
+              Entity prevEntity = valueDifference.leftValue();
+              Entity currEntity = valueDifference.rightValue();
+              sendUpdateNotification(prevEntity, currEntity);
+            });
+
+    mapDifference
+        .entriesOnlyOnLeft()
+        .entrySet()
+        .forEach(
+            entry -> {
+              sendDeleteNotification(entry.getValue());
+            });
+  }
+
+  private void sendCreateNotification(Entity createdEntity) {
+    Builder builder = EntityChangeEventValue.newBuilder();
+    builder.setCreateEvent(EntityCreateEvent.newBuilder().setCreatedEntity(createdEntity).build());
+    entityChangeEventProducer.send(getEntityChangeEventKey(createdEntity), builder.build());
+  }
+
+  private void sendUpdateNotification(Entity prevEntity, Entity currEntity) {
+    Builder builder = EntityChangeEventValue.newBuilder();
+    builder.setUpdateEvent(
+        EntityUpdateEvent.newBuilder()
+            .setPreviousEntity(prevEntity)
+            .setLatestEntity(currEntity)
+            .build());
+    entityChangeEventProducer.send(getEntityChangeEventKey(currEntity), builder.build());
+  }
+
+  private void sendDeleteNotification(Entity deletedEntity) {
+    Builder builder = EntityChangeEventValue.newBuilder();
+    builder.setDeleteEvent(EntityDeleteEvent.newBuilder().setDeletedEntity(deletedEntity).build());
+    entityChangeEventProducer.send(getEntityChangeEventKey(deletedEntity), builder.build());
+  }
+
+  private EntityChangeEventKey getEntityChangeEventKey(Entity entity) {
```

Comment:
discussed briefly last week, but may need to serialize something else for kafka keys given the protobuf non-determinism.

Reply:
A string concatenated can be used

Reply:
this seems like an important one to follow up on one way or the other - we don't want to change the key later.

Reply:
> Entity id is random enough to distribute well across partition. Also entity type and tenant id are available as part of entity itself.

This depends on the entity type. If there's no penalty (the consumer doesn't care what the key is - right?), I'd go for the concatenated key. For V1 entities, the entity ID is hashed across identifying attributes, type and tenant id so should be globally unique and fine, but for V2 entities, the entity ID is opaque to entity service and determined by the definition. The uniqueness across types and tenant IDs for v2 entities is treated as an internal impl detail and is only accounted for at the doc key. What this means is, for example, imagine a v2 service entity (service is currently v1, but it explains the general issue for v2 entities) which is identified by the servicename attribute. Across tenants, the name "data-service" or "iam" for example, may be common resulting in identical user-facing entityIds (but unique doc ids).

Is it going to be random enough if we just use the entity id? I've got to imagine yes, and I'm fine with that if there's any real downside. It just seems like cheap insurance to concatenate the other information, too (for the record, that's exactly how the doc key is generated for v2 entities, too - `String.format("%s:%s:%s", this.tenantId, this.entityType, this.entityId)`).

Reply:
Used logic similar to doc key to generate the key